draft-yasskin-wpack-bundled-exchanges.md

```diff
@@ -289,103 +339,130 @@ steps, taking the `stream` as input.
       Note: The `ignoredSections` enables sections that supercede other sections
       to be introduced in the future. Implementations that don't implement any
       such sections are free to omit the relevant steps.
-   1. If `sectionOffsets["name"]` exists, return an error. That is, duplicate
-      sections are forbidden.
+   1. If `sectionOffsets["name"]` exists, return a "format error". That is,
+      duplicate sections are forbidden.
    1. Set `sectionOffsets["name"]` to (`currentOffset`, `length`).
    1. Set `currentOffset` to `currentOffset + length`.
 
-1. If the "responses" section is not last in `sectionLengths`, return an error.
-   This allows a streaming parser to assume that it'll know the requests by the
-   time their responses arrive.
+1. If the "responses" section is not last in `sectionLengths`, return a "format
+   error" with `fallbackUrl`. This allows a streaming parser to assume that
+   it'll know the requests by the time their responses arrive.
 
-1. Let `metadata` be an empty map ({{INFRA}}).
+1. Let `metadata` be a map ({{INFRA}}) initially containing the single key/value
+   pair `"primaryUrl"`/`fallbackUrl`.
 
 1. For each `"name"` → (`offset`, `length`) triple in `sectionOffsets`:
    1. If `"name"` isn't in `knownSections`, continue to the next triple.
    1. If `"name"`'s Metadata field ({{section-name-registry}}) is "No", continue
       to the next triple.
    1. If `"name"` is in `ignoredSections`, continue to the next triple.
-   1. Seek to offset `offset` in `stream`. If this fails, return an error.
+   1. Seek to offset `offset` in `stream`. If this fails, return a "format
+      error" with `fallbackUrl`.
    1. Let `sectionContents` be the result of reading `length` bytes from
-      `stream`. If `sectionContents` is an error, return that error.
+      `stream`. If `sectionContents` is an error, return a "format error" with
+      `fallbackUrl`.
    1. Follow `"name"`'s specification from `knownSections` to process the
       section, passing `sectionContents`, `stream`, `sectionOffsets`, and
-      `metadata`. If this returns an error, return it.
+      `metadata`. If this returns an error, return a "format error" with
+      `fallbackUrl`.
 
-1. If `metadata` doesn't have entries with keys "requests" and "manifest",
-   return an error.
+1. If `metadata` doesn't have entries with keys "primaryUrl", "requests" and
+   "manifest", return a "format error" with `fallbackUrl`.
```

Comment:
At this point we must have primaryUrl (or we can't return an error with fallbackUrl)?

Reply:
Yep, I've moved this to an Assert on the previous line.